Packages/com.unity.multiplayer.samples.coop/Utilities/ServerAdditiveSceneLoader.cs

```diff
@@ -54,58 +60,54 @@ public override void OnNetworkDespawn()
         {
-            if (!IsSpawned)
+            if (other.CompareTag(playerTag) && other.TryGetComponent(out NetworkObject networkObject))
             {
-                return;
-            }
+                m_PlayersInTrigger.Add(networkObject.OwnerClientId);
 
-            if (!m_IsCooldown)
-            {
-                if (m_IsLoaded && m_PlayersInTrigger.Count == 0)
+                if (m_UnloadCoroutine != null)
                 {
-                    NetworkManager.SceneManager.UnloadScene(SceneManager.GetSceneByName(sceneName));
-                    m_IsLoaded = false;
+                    // stopping the unloading coroutine since there is now a player-owned NetworkObject inside
+                    StopCoroutine(m_UnloadCoroutine);
                 }
-                else if (!m_IsLoaded && m_PlayersInTrigger.Count > 0)
+
+                if (!m_IsLoaded && m_PlayersInTrigger.Count > 0)
                 {
                     NetworkManager.SceneManager.LoadScene(sceneName, LoadSceneMode.Additive);
                     m_IsLoaded = true;
-
-                    // Add this delay to prevent players entering and leaving the collider repeatedly from continually load/unloading the scene
-                    StartCoroutine(Cooldown());
                 }
             }
         }
 
-        void OnTriggerEnter(Collider other)
-        {
-            if (other.CompareTag(playerTag) && other.TryGetComponent(out NetworkObject networkObject))
-            {
-                m_PlayersInTrigger.Add(networkObject.OwnerClientId);
-            }
-        }
-
         void OnTriggerExit(Collider other)
         {
             if (other.CompareTag(playerTag) && other.TryGetComponent(out NetworkObject networkObject))
             {
                 m_PlayersInTrigger.Remove(networkObject.OwnerClientId);
+                if (m_IsLoaded && m_PlayersInTrigger.Count == 0)
+                {
+                    // using a coroutine here to add a delay before unloading the scene
+                    m_UnloadCoroutine = StartCoroutine(UnloadCoroutine());
+                }
             }
         }
 
         void RemovePlayer(ulong clientId)
         {
-            // remove all references to this clientId
+            // remove all references to this clientId. There could be multiple references if a single client owns
+            // multiple NetworkObjects with the playerTag, or if this script's GameObject has overlapping colliders
             while (m_PlayersInTrigger.Remove(clientId)) { }
         }
 
-        IEnumerator Cooldown()
+        IEnumerator UnloadCoroutine()
         {
-            m_IsCooldown = true;
-            yield return new WaitForSeconds(cooldownBeforeUnload);
-            m_IsCooldown = false;
+            yield return new WaitForSeconds(delayBeforeUnload);
+            if (m_IsLoaded)
+            {
+                NetworkManager.SceneManager.UnloadScene(SceneManager.GetSceneByName(sceneName));
```

Comment:
UnloadScene is async though, can we really say m_IsLoaded = false right after calling this?

Reply:
It's an issue with the field's name. The intent here is simply to prevent asking to load/unload the same scene more than once. I'll rename it to make it clearer.
However it highlights another issue I did not see before: if a scene takes longer to load than the `delayBeforeUnload` it would not be able to unload it. I'll think of a way to deal with this edge-case.